File: login-server/src/main/java/work/fking/pangya/login/task/LoginTask.java
package work.fking.pangya.login.task;

import io.netty.channel.Channel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import work.fking.pangya.discovery.ServerType;
import work.fking.pangya.login.LoginServer;
import work.fking.pangya.login.auth.UserInfo;
import work.fking.pangya.login.net.ClientPacketDispatcher;
import work.fking.pangya.login.net.ClientPacketType;
import work.fking.pangya.login.net.ClientProtocol;
import work.fking.pangya.login.net.GameProtocolDecoder;
import work.fking.pangya.login.net.LoginState;
import work.fking.pangya.login.packet.outbound.LoginReplies;
import work.fking.pangya.login.packet.outbound.LoginReplies.Error;
import work.fking.pangya.login.packet.outbound.ServerListReplies;

public class LoginTask implements Runnable {

    private static final ClientProtocol PROTOCOL = ClientProtocol.create(ClientPacketType.values());
    private static final Logger LOGGER = LoggerFactory.getLogger(LoginTask.class);

    private final LoginServer loginServer;
    private final Channel channel;
    private final int cryptKey;
    private final String username;
    private final char[] password;

    public LoginTask(LoginServer loginServer, Channel channel, int cryptKey, String username, char[] password) {
        this.loginServer = loginServer;
        this.channel = channel;
        this.cryptKey = cryptKey;
        this.username = username;
        this.password = password;
    }

    @Override
    public void run() {
        UserInfo userInfo;
        try {
            userInfo = loginServer.authenticator().authenticate(username, password);
        } catch (Exception e) {
            // TODO: improve this
            LOGGER.warn("Authentication failed", e);
            channel.writeAndFlush(LoginReplies.error(Error.INCORRECT_USERNAME_PASSWORD));
            return;
        }
        var player = loginServer.registerPlayer(channel, userInfo);

        var discoveryClient = loginServer.discoveryClient();
        var gameServers = discoveryClient.instances(ServerType.GAME);
        var socialServers = discoveryClient.instances(ServerType.SOCIAL);

        var pipeline = channel.pipeline();
        pipeline.remove("loginHandler");
        pipeline.addLast("decoder", GameProtocolDecoder.create(PROTOCOL, cryptKey));
        pipeline.addLast("packetDispatcher", ClientPacketDispatcher.create(loginServer, player, PROTOCOL.handlers()));

        channel.write(LoginReplies.loginKey(player.loginKey()));
        channel.write(LoginReplies.chatMacros());
        channel.write(LoginReplies.success(player.uid(), player.username(), player.nickname()));
        channel.write(ServerListReplies.gameServers(gameServers));
        channel.write(ServerListReplies.socialServers(socialServers));
        player.setLoginState(LoginState.LOGGED_IN);
        channel.flush();
    }
}

File: login-server/src/main/kotlin/work/fking/pangya/login/task/LoginTask.kt
package work.fking.pangya.login.task

import io.netty.channel.Channel
import org.slf4j.LoggerFactory
import work.fking.pangya.login.LoginServer
import work.fking.pangya.login.auth.ExceptionResult
import work.fking.pangya.login.auth.InvalidCredentialsResult
import work.fking.pangya.login.auth.SuccessResult
import work.fking.pangya.login.auth.UserInfo
import work.fking.pangya.login.net.ClientPacketDispatcher
import work.fking.pangya.login.net.ClientPacketType
import work.fking.pangya.login.net.ClientProtocol
import work.fking.pangya.login.net.LoginState.LOGGED_IN
import work.fking.pangya.login.net.LoginState.SELECTING_CHARACTER
import work.fking.pangya.login.net.LoginState.SELECTING_NICKNAME
import work.fking.pangya.login.net.pipe.ProtocolDecoder
import work.fking.pangya.login.packet.outbound.LoginReplies
import work.fking.pangya.login.packet.outbound.LoginReplies.Error.INCORRECT_USERNAME_PASSWORD
import work.fking.pangya.login.packet.outbound.LoginReplies.Error.INVALID_ID
import work.fking.pangya.login.packet.outbound.LoginReplies.Error.SERVER_MAINTENANCE
import work.fking.pangya.login.session.Active
import work.fking.pangya.login.session.Error
import work.fking.pangya.login.session.NotFound
import work.fking.pangya.login.session.Timeout

private val PROTOCOL: ClientProtocol = ClientProtocol(ClientPacketType.entries.toTypedArray())
private val LOGGER = LoggerFactory.getLogger(LoginTask::class.java)

class LoginTask(
    private val loginServer: LoginServer,
    private val channel: Channel,
    private val cryptKey: Int,
    private val username: String,
    private val password: ByteArray
) : Runnable {

    override fun run() {
        when (val authResult = loginServer.authenticator.authenticate(username, password)) {
            is SuccessResult -> onSuccessAuth(authResult.userInfo)
            is InvalidCredentialsResult -> onInvalidCredentials()
            is ExceptionResult -> LOGGER.warn("Auth failed", authResult.exception)
        }
    }

    private fun onSuccessAuth(userInfo: UserInfo) {
        val sessionClient = loginServer.sessionClient
        val result = sessionClient.findSessionByUsername(username)

        when (result) {
            is Active -> duplicateConnection()
            is NotFound, Timeout -> proceedLogin(userInfo)
            is Error -> serverError()
        }
    }

    private fun proceedLogin(userInfo: UserInfo) {
        val player = loginServer.registerPlayer(channel, userInfo)
        loginServer.sessionClient.registerSession(player)

        val pipeline = channel.pipeline()
        pipeline.remove("loginHandler")
        pipeline.remove("timeoutHandler")
        pipeline.addLast("decoder", ProtocolDecoder(PROTOCOL, cryptKey))
        pipeline.addLast("packetDispatcher", ClientPacketDispatcher(loginServer, player, PROTOCOL.handlers()))

        val nickname = player.nickname
        player.state = if (nickname == null) {
            channel.write(LoginReplies.createNickname())
            SELECTING_NICKNAME
        } else if (!player.hasBaseCharacter) {
            channel.write(LoginReplies.selectCharacter())
            SELECTING_CHARACTER
        } else {
            loginServer.proceedPlayerToLoggedIn(player)
            LOGGED_IN
        }
        channel.flush()
    }

    private fun onInvalidCredentials() {
        channel.writeAndFlush(LoginReplies.error(INCORRECT_USERNAME_PASSWORD))
    }

    private fun duplicateConnection() {
        channel.writeAndFlush(LoginReplies.error(INVALID_ID, "Account is already logged in."))
    }

    private fun serverError() {
        channel.writeAndFlush(LoginReplies.error(SERVER_MAINTENANCE))

    }
}
